Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt

```diff
@@ -48,4 +57,71 @@ class SimperiumTagsRepository(
             false
         }
     }
+
+    override suspend fun deleteTag(tag: Tag) = withContext(Dispatchers.IO) {
+        deleteTagFromNotes(tag)
+        tag.delete()
+    }
+
+    private fun deleteTagFromNotes(tag: Tag) {
+        val cursor = tag.findNotes(notesBucket, tag.name)
+
+        while (cursor.moveToNext()) {
+            val note = cursor.getObject()
+            val tags = note.tags
+            tags.remove(tag.name)
+            note.tags = tags
+            note.save()
+        }
+
+        cursor.close()
+    }
+
+    @ExperimentalCoroutinesApi
+    override suspend fun tagsChanged(): Flow<Boolean> = callbackFlow {
+        val callbackOnSaveObject = Bucket.OnSaveObjectListener<Tag> { _, _ -> offer(true) }
+        val callbackOnDeleteObject = Bucket.OnDeleteObjectListener<Tag> { _, _ -> offer(true) }
+        val callbackOnNetworkChange = Bucket.OnNetworkChangeListener<Tag> { _, _, _ -> offer(true) }
+
+        tagsBucket.addOnSaveObjectListener(callbackOnSaveObject)
+        tagsBucket.addOnDeleteObjectListener(callbackOnDeleteObject)
+        tagsBucket.addOnNetworkChangeListener(callbackOnNetworkChange)
+        AppLog.add(AppLog.Type.SYNC, "Added tag bucket listener (TagsActivity)")
+
+        awaitClose {
+            tagsBucket.removeOnSaveObjectListener(callbackOnSaveObject)
+            tagsBucket.removeOnDeleteObjectListener(callbackOnDeleteObject)
+            tagsBucket.removeOnNetworkChangeListener(callbackOnNetworkChange)
+            AppLog.add(AppLog.Type.SYNC, "Removed tag bucket listener (TagsActivity)")
+        }
+    }
+
+    override suspend fun allTags(): List<TagItem> = withContext(Dispatchers.IO) {
+        val tagQuery = Tag.all(tagsBucket).reorder().orderByKey().include(Tag.NOTE_COUNT_INDEX_NAME)
+        val cursor = tagQuery.execute()
+
+        return@withContext cursorToTagItems(cursor)
+    }
+
+    override suspend fun searchTags(query: String): List<TagItem> = withContext(Dispatchers.IO) {
+        val tags = Tag.all(tagsBucket)
+                .where(Tag.NAME_PROPERTY, Query.ComparisonType.LIKE, "%$query%")
+                .orderByKey().include(Tag.NOTE_COUNT_INDEX_NAME)
+                .reorder()
+        val cursor = tags.execute()
+
+        return@withContext cursorToTagItems(cursor)
+    }
+
+    private fun cursorToTagItems(cursor: Bucket.ObjectCursor<Tag>): List<TagItem> {
+        return (1 .. cursor.count).map {
+            cursor.moveToNext()
+            val tag = cursor.`object`
+            val noteCount: Int = notesBucket
+                    .query()
+                    .where(Note.TAGS_PROPERTY, Query.ComparisonType.EQUAL_TO, tag.name)
```

Comment:
unfortunately this is like the other places and we probably can't use the database query directly. if we don't compare tag hashes we will get the wrong counts. this has been an annoying bug in the app for a long time, so it's not like it's new, but also, we should avoid introducing new regressions.

Reply:
@dmsnell thanks for the feedback. It is clear that there are some bugs regarding tag names and how we compare them and search them. All these queries are passed directly from the original activity to keep the same behavior while refactoring. Do you think we should make all these changes in this PR or make separate issues to introduce the changes?

Reply:
> All these queries were passed directly (copied) from the original activity to keep the same behavior while refactoring

Ugh. Bummer - sorry you have to deal with this!

> Do you think we should make all these changes in this PR or make separate issues to introduce the changes?

Like I said, this app has a lot of issues with tags. A separate PR is fine. Would it be much work to propose that change _before_ this refactor? It doesn't matter too much, but I prefer seeing that we fix something before moving it around to another place, leaving its trail of buggy code in the project history.

Reply:
> Ugh. Bummer - sorry you have to deal with this! 
No problem 😅️. Let's fix them all!

> It doesn't matter too much, but I prefer seeing that we fix something before moving it around to another place, leaving its trail of buggy code in the project history.

We could do it before on a separate PR, although I am inclining on doing the changes in this PR since we have the code encapsulated in the SimperiumTagsRepository and the changes might not be that big (few lines of code). 

What do you think about fixing the search problem and the delete from notes problem in this PR?